app/models/concerns/foreman_bootdisk/orchestration/compute.rb

```diff
@@ -46,6 +47,16 @@ def bootdisk_attach_iso
         compute_resource.iso_attach(File.basename(bootdisk_isofile), uuid)
       end
 
+      def bootdisk_detach_iso
+        return true unless provision_method == 'bootdisk'
+        return true if previous_changes['build'].nil?
+
+        if previous_changes['build'] == [true, false]
```

Comment:
I don't quite get this, shouldn't `bootdisk_detach_iso` only run `before_provision` anyway?

Why don't you make this part of the orchestration queue?
E.g. modifying `queue_bootdisk_compute` to queue a detach task when the host leaves build mode? There should be [plenty of examples](https://github.com/theforeman/foreman/blob/c86ed9c68adedaaff5562ddaa6e7229ec416427d/app/models/concerns/orchestration/dhcp.rb#L156) that demonstrate the required code changes.

Reply:
regarding orchestration queue: detaching should be done before starting the provisioned system. I guess, its not possible to add a task right before that stage. "before provision" is executed after the host has signalized the "built" flag (in the build snippet). Therefore the procedure should be:
- in provisioning template, the host "eject" the cdrom
- in provisioning template the snippet build runs and sends the built flag
- bootdisk_detach_iso is used to finally remove the ISO

Reply:
When the host requests the built callback, the queue also runs. I believe this is used e.g. for tftp orchestration to switch the host from pxe boot to local boot.

Reply:
I guess you mean this one: https://github.com/theforeman/foreman/blob/bad5f878b6fb275dbc7267930bcd0cb589d9aa66/app/models/concerns/orchestration.rb#L13 This is actually run to the wrong point in time.. :(

Reply:
No, I mean the default queue.

Reply:
I can not find out how to specify when a task should be run when using the queue. Of course, there are priorities but this isn't really "safe".

Reply:
As Timo said, the best place to learn would be TFTP orchestration when Foreman flips TFTP config back to boot from local drive. They are hooked via `after_validation :queue_tftp` and then it compares the `built` flag. If it was updated, it schedules redeployment of files. Isn't this exactly what you are looking for?